src/main/java/com/direwolf20/buildinggadgets/common/schema/template/Template.java

```diff
@@ -0,0 +1,241 @@
+        bounds = new BoundingBox(minX, minY, minZ, maxX, maxY, maxZ);
+
+        return Optional.of(new Template(states, bounds, author));
+    }
+
+    private static int getId(long packed) {
+        return (int) ((packed >> 40) & BYTE_MASK_24BIT);
+    }
+
+    private static long getPos(long packed) {
+        return packed & BYTE_MASK_40BIT;
+    }
+
+    private Template(Long2ObjectMap<BlockState> states, BoundingBox bounds, String author) {
+        this.states = states;
+        this.bounds = bounds;
+        this.author = author;
+    }
+
+    public BoundingBox getBounds() {
+        return bounds;
+    }
+
+    public String getAuthor() {
+        return author;
+    }
+
+    public int size() {
+        return states.size();
+    }
+
+    /**
+     * Converts a {@link Vec3i} into a long representation. Notice that only the first 40bits are used, because the
+     * remaining 24bit are reserved for state-id's. This is also why {@link BlockPos#toLong()} cannot be used.
+     *
+     * @param vec The vector to "longify"
+     * @return A long consisting of [24bit-unused]|[14bit-x_coord]|[12bit-y_coord]|[14bit-z_coord]
+     */
+    private long vecToLong(Vec3i vec) {
+        long res = (long) (vec.getX() & BYTE_MASK_14BIT) << 26;
+        res |= (vec.getY() & BYTE_MASK_12BIT) << 14;
+        res |= (vec.getZ() & BYTE_MASK_14BIT);
+        return res;
+    }
+
+    /**
+     * Converts longs provided by {@link #vecToLong(Vec3i)} back into a {@link Mutable MutableBlockPos}.
+     *
+     * @param setTo      the {@link Mutable} to apply values to
+     * @param serialized the long produced by {@link #vecToLong(Vec3i)}
+     * @return {@code setTo} to allow for Method chainging.
+     */
+    private static Mutable posFromLong(Mutable setTo, long serialized) {
+        int x = (int) ((serialized >> 26) & BYTE_MASK_14BIT);
+        int y = (int) ((serialized >> 14) & BYTE_MASK_12BIT);
+        int z = (int) (serialized & BYTE_MASK_14BIT);
+        setTo.setPos(x, y, z);
+        return setTo;
+    }
+
+    private long includeId(long posLong, int stateId) {
+        return posLong | ((long) (stateId & BYTE_MASK_24BIT) << 40);
+    }
+
+    /**
+     * Warning: do not store the {@link TemplateData} without invoking {@link TemplateData#copy()} first.
+     * This implementation uses only a single {@link TemplateData} and a single {@link Mutable} instance during the
+     * iteration which are mutated to fit to the current iteration point. Therefore a new invocation of
+     * {@link Iterator#next()} will reset the data viewed by a reference to this {@link Iterator}'s result.
+     *
+     * This implementation provides {@code O(1)} Object allocations during during iteration (assuming no tile data
+     * to be present) in order to be usable with large Templates without over-utilising the garbage collector.
+     * For an idea of the performance of this mutable variant compared to recreating the Object in each step,
```

Comment:
This is over the top commenting.

Reply:
And then someone will ask why this is done, the way it is done. This is going to stay.